third_party/move/move-compiler-v2/src/env_pipeline/function_checker.rs

```diff
@@ -3,55 +3,190 @@
 
 //! Do a few checks of functions and function calls.
 
-use crate::Options;
+use crate::{experiments::Experiment, Options};
 use codespan_reporting::diagnostic::Severity;
 use move_binary_format::file_format::Visibility;
 use move_model::{
     ast::{ExpData, Operation, Pattern},
-    model::{FunId, FunctionEnv, GlobalEnv, Loc, ModuleEnv, NodeId, QualifiedId},
+    model::{FunId, FunctionEnv, GlobalEnv, Loc, ModuleEnv, NodeId, Parameter, QualifiedId},
     ty::Type,
 };
-use std::{collections::BTreeSet, iter::Iterator, vec::Vec};
+use std::{collections::BTreeSet, iter::Iterator, ops::Deref, vec::Vec};
 
 type QualifiedFunId = QualifiedId<FunId>;
 
-/// check that non-inline function parameters do not have function type.
+fn type_has_function(ty: &Type) -> bool {
+    match ty {
+        Type::Tuple(tys) => tys.iter().any(|ty| ty.is_function()),
+        Type::Fun(..) => true,
+        _ => false,
+    }
+}
+
+// Takes a list of function types, returns those which have a function type in their argument type
+fn identify_function_types_with_functions_in_args(func_returns: Vec<Type>) -> Vec<Type> {
```

Comment:
Because this function a list of function types, could we call the param `func_types` or even just `types`? Not clear from the local context why this param is called `func_returns`.